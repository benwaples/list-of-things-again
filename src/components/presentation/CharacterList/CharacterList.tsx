import React from 'react'
import { useCharacters } from '../../../hooks'
import { Character } from '../Character/Character'

type CharacterProps = {
  id: string, 
  name: string, 
  imageUrl: string
}

export const CharacterList = () => {
  const { characters, loading, error } = useCharacters()

  const charactersElements = characters.map((character: CharacterProps) => <li key={character.id}>
        <Character {...character}/>
      </li>
  )

  if(loading) return <h1>Loading...</h1>
  if(error) return <h1>failed to get characters</h1>
  return (
    <div>
      {charactersElements}
    </div>
  )
}
