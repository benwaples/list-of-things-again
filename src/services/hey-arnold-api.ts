type CharacterShape = {
  _id: string, 
  name: string, 
  image: string
}

export type Character = {
  id: string, 
  name: string, 
  imageUrl: string
}

export const getCharacters = (): Promise<Character[]> => {
  return fetch('https://hey-arnold-api.herokuapp.com/api/v1/characters')
    .then(res => res.json())
    .then(json => json.map((character: CharacterShape) => ({
      id: character._id,
      name: character.name,
      imageUrl: character.image
    })))
}

export const getCharacterById = (id: string): Promise<Character> => {
  return fetch(`https://hey-arnold-api.herokuapp.com/api/v1/characters/${id}`)
    .then(res => res.json())
    .then((character: CharacterShape) => ({
      id: character._id,
      name: character.name,
      imageUrl: character.image
    }))
}