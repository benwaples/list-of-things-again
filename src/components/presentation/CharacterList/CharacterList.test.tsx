import React from 'react'
import {
  screen,
  waitFor,
  render
} from '@testing-library/react'
import { getCharacters } from '../../../services/hey-arnold-api'
import { MemoryRouter } from 'react-router-dom'
import { CharacterList } from './CharacterList'
import { act } from 'react-dom/test-utils'

jest.mock('../../../services/hey-arnold-api')

describe('CharacterList component behavior test', () => {

  it('should render a list of characters', () => {
   act(() => {

     getCharacters.mockResolvedValue([
       {
         id: 'my', 
         name: 'name', 
         imageUrl: 'is'
       }
   ])
  })

    render(
      <MemoryRouter>
        <CharacterList page={1} />
      </MemoryRouter>
    )
  
    screen.getByText('Loading...')
  

    return waitFor(() => {
      expect(screen.getByText('name')).not.toBeEmptyDOMElement()
    })
  })

  it('CharacterList snapshot', () => {
    act(() => {
      getCharacters.mockResolvedValue([
        {
          id: 'my', 
          name: 'name', 
          imageUrl: 'is'
        }
    ])
   })
    
    const { asFragment } = render(
      <MemoryRouter>
        <CharacterList page={1}/>
      </MemoryRouter>
    )

    return waitFor(() => {
      // eslint-disable-next-line testing-library/no-wait-for-snapshot
      expect(asFragment()).toMatchSnapshot()
    })
  })
  
})
