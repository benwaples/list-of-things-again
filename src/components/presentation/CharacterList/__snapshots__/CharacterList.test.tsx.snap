// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList component behavior test CharacterList snapshot 1`] = `
<DocumentFragment>
  <h1>
    Loading...
  </h1>
</DocumentFragment>
`;

exports[`CharacterList component behavior test CharacterList snapshot 2`] = `
<DocumentFragment>
  <div>
    <li>
      <div>
        <figure>
          <img
            alt="name"
            src="is"
          />
          <figcaption>
            name
          </figcaption>
        </figure>
        <a
          href="/characters/my"
        >
          See More
        </a>
      </div>
    </li>
  </div>
</DocumentFragment>
`;
